Make kernel start in project dir root default and add setting

buyology opened the pull request:
xref [#523](https://github.com/nteract/hydrogen/issues/523#issuecomment-300893935)

* Makes project directory root the default kernel start directory
* Adds setting (`startInProjectRootDir`) to control this behaviour with the possibility to revert to previous behaviour (start kernel in same dir as file current file)

lgeiger commented on lib/kernel-manager.js:
```diff
@@ -86,11 +86,18 @@ class KernelManager {
     log("KernelManager: startKernel:", displayName);
 
+    if (atom.config.get("Hydrogen.startInProjectRootDir")) {
+      kernelStartDir = atom.project.getPaths()[0];
```
I think `atom.project.getPaths()` can be empty if no project is present in the tree view. In that case we should fall back to the `getEditorDirectory(store.editor)`.

A very nice touch (now or in a future PR) would be to use the project where the current file belongs to and not just the first one. This could be handy if multiple projects are open.

BenRussert commented on lib/config.js:
```diff
@@ -28,6 +28,13 @@ const Config = {
+    startInProjectRootDir: {
+      title: "Start kernel in project root directory",
+      includeTitle: false,
+      description: "If enabled, the kernel is started in the project's root directory (default). If disabled, it is started in the current file's directory. Restart Atom for changes to take effect.",
```
> Restart Atom for changes to take effect.

I think you could just restart the kernel, right?

buyology commented:
* Added check for length of `atom.project.getPaths()`-array
* Added support for multiprojects set-up (didn't know you could do this!)
* Refactored the settings into an `enum` with these three settings:
  * `firstProjectDir`
  * `projectDirOfFile`
  * `dirOfFile`

lgeiger commented on lib/kernel-manager.js:
```diff
@@ -86,11 +86,26 @@ class KernelManager {
 
+    switch (atom.config.get("Hydrogen.startDir")) {
+      case "firstProjectDir":
+        if (atom.project.getPaths().length > 0) {
```
If this condition isn't fulfilled it falls back to `case "projectDirOfFile":` which will return `/`. I think it should fall back to `currentPath` instead.

buyology replied:
Agreed!

lgeiger commented on lib/kernel-manager.js:
```diff
@@ -86,11 +86,26 @@ class KernelManager {
+          break;
+        }
+      case "projectDirOfFile":
+        kernelStartDir = atom.project.relativizePath(currentPath)[0];
```
Using `atom.project.relativizePath` can be a bit problematic since it will return `/` if no project is found or if there isn't a project related to the current file. I think it should fall back to `currentPath` instead of `/`. What do you think?

buyology replied:
Was just going to write and say we need to catch this :) According to the docs it should return `null` if there is no projectpath? Need to test this. 

Is the way to test "without" a project path to open a file directly in atom?

lgeiger replied:
> Is the way to test "without" a project path to open a file directly in atom?

You can remove the project directory from the tree view via the context menu.

buyology commented:
updated

lgeiger commented:
Thank you for the work! 👍 

@buyology Would you like to help us maintain this package?  😄

lgeiger commented:
On thing that would be super cool is to move this logic to a function in `utils.js` and add a few tests, so we can reuse it in the future.
